src/Workspaces/Core/Portable/Serialization/SerializerService.cs

```diff
@@ -174,6 +183,74 @@ public void Serialize(object value, ObjectWriter writer, CancellationToken cance
         }
     }
 
+    private static void Write(ObjectWriter writer, ImmutableDictionary<string, StructuredAnalyzerConfigOptions> optionsByLanguage)
+    {
+        // Only serialize options for C#/VB since other languages are not OOP.
+
+        var csOptions = ImmutableDictionary.GetValueOrDefault(optionsByLanguage, LanguageNames.CSharp);
+        var vbOptions = ImmutableDictionary.GetValueOrDefault(optionsByLanguage, LanguageNames.VisualBasic);
+
+        writer.WriteCompressedUInt((uint)((csOptions != null ? 1 : 0) + (vbOptions != null ? 1 : 0)));
+
+        Serialize(LanguageNames.CSharp, csOptions);
+        Serialize(LanguageNames.VisualBasic, vbOptions);
+
+        void Serialize(string language, StructuredAnalyzerConfigOptions? options)
+        {
+            if (options != null)
+            {
+                writer.WriteString(language);
+
+                // order for deterministic checksums
+                foreach (var key in options.Keys.Order())
+                {
+                    if (options.TryGetValue(key, out var value))
+                    {
+                        writer.WriteString(key);
+                        writer.WriteString(value);
+                    }
+                }
+
+                // terminator:
+                writer.WriteString(null);
```

Comment:
i feel oogy about this.  options.Keys should have a known Count.  so we can write that out first, and have the reader read that in and read the right amount of values.

while i can't think of anything that would break with this current code... i really feel certain it will bite us in the ass at some point :)

Reply:
option.Keys is IEnumerable - does not have Count. I can enumerate to determine the count and then again, but seems unnecessary to do so. NULL-termination works fine imo.